src/rtc_stream.cpp

```diff
@@ -0,0 +1,298 @@
+	auto addr = m_peer_connection->localAddress();
+	if(!addr)
+	{
+		ec = boost::asio::error::operation_not_supported;
+		return endpoint_type();
+	}
+
+	size_t pos = addr->find_last_of(':');
+	if(pos == std::string::npos)
+	{
+		ec = boost::asio::error::address_family_not_supported;
+		return endpoint_type();
+	}
+
+	return endpoint_type(ip::make_address(addr->substr(0, pos), ec)
+			, std::stoul(addr->substr(pos+1)));
+}
+
+void rtc_stream::add_read_buffer(void* buf, std::size_t const len)
+{
+    TORRENT_ASSERT(len < INT_MAX);
+    TORRENT_ASSERT(len > 0);
+    TORRENT_ASSERT(buf);
+
+    m_read_buffer.emplace_back(buf, len);
+    m_read_buffer_size += len;
+}
+
+void rtc_stream::add_write_buffer(void const* buf, std::size_t const len)
+{
+    TORRENT_ASSERT(len < INT_MAX);
+    TORRENT_ASSERT(len > 0);
+    TORRENT_ASSERT(buf);
+
+    m_write_buffer.emplace_back(const_cast<void*>(buf), len);
+	m_write_buffer_size += len;
+}
+
+void rtc_stream::issue_read()
+{
+	TORRENT_ASSERT(m_read_handler);
+	TORRENT_ASSERT(m_read_buffer_size > 0);
+
+	if(!ensure_open()) return;
+
+	std::size_t bytes_read = read_some(false);
+	if(bytes_read > 0)
+	{
+		post(m_io_context, std::bind(m_read_handler, error_code{}, bytes_read));
+		m_read_handler = nullptr;
+
+		m_read_buffer_size = 0;
+        m_read_buffer.clear();
+	}
+}
+
+std::size_t rtc_stream::read_some(bool const clear_buffers)
+{
+	if(!ensure_open()) return 0;
+
+	std::size_t ret = 0;
+	auto target = m_read_buffer.begin();
+	while(!m_incoming.empty() && target != m_read_buffer.end())
+	{
+		auto& message = m_incoming.front();
+		std::size_t to_copy = std::min(message.size(), target->len);
+		std::memcpy(target->buf, message.data(), to_copy);
+		ret += to_copy;
+		target->buf = static_cast<char*>(target->buf) + to_copy;
+		TORRENT_ASSERT(target->len >= to_copy);
+		target->len -= to_copy;
+		TORRENT_ASSERT(m_read_buffer_size >= to_copy);
+		m_read_buffer_size -= to_copy;
+
+		// Move to next target
+		if (target->len == 0) target = m_read_buffer.erase(target);
+
+		if (to_copy == message.size())
+		{
+			// Consumed entire message
+			m_incoming.pop();
+		}
+		else {
+			message.erase(message.begin(), message.begin() + to_copy);
+		}
+
+		TORRENT_ASSERT(m_incoming_size >= to_copy);
+		m_incoming_size -= to_copy;
+	}
+
+	if (clear_buffers)
+	{
+		m_read_buffer_size = 0;
+		m_read_buffer.clear();
+	}
+	return ret;
+}
+
+void rtc_stream::issue_write()
+{
+	TORRENT_ASSERT(m_write_handler);
+	TORRENT_ASSERT(m_write_buffer_size > 0);
+
+	if(!ensure_open()) return;
+
+	std::size_t bytes_written = 0;
+	auto target = m_write_buffer.begin();
+	while(target != m_write_buffer.end())
+	{
+		m_data_channel->send(static_cast<rtc::byte const*>(target->buf), target->len);
+		bytes_written += target->len;
+		TORRENT_ASSERT(m_write_buffer_size >= target->len);
+		m_write_buffer_size -= target->len;
+		target = m_write_buffer.erase(target);
+	}
+
+	post(m_io_context, std::bind(m_write_handler, error_code{}, bytes_written));
```

Comment:
this will post the handler right away. Does the datachannel not support waiting for the socket to become writable? does it buffer indefinitely under the hood?
If I'm uploading a 10GiB file over a slow connection, will this just end up allocating 10GiB of RAM?

Reply:
Yes you're right, that's a similar issue than when reading. On Web, there is a dubious API allowing to watch the buffer level, but its implementation is somewhat inconsistent. I'll see what I can do here.

Reply:
This should be solved with the custom callback I added to the data channels.